w3d_translator/read_xml.py

```diff
@@ -1,50 +1,97 @@
     # TODO: Build each <Timeline> in <TimelineRoot> (8)
-    # timeline_root = {}
-    for tag in root.find("TimelineRoot") or []:
-        pass  # Dict of Timeline by name
+    return story
 
-    # TODO: Build each <Placement> in <PlacementRoot> (9)
-    # placement_root = {}
-    for tag in root.find("PlacementRoot") or []:
-        pass  # Dict of Placement by name
 
-    # TODO: Build each <Sound> in <SoundRoot> (10)
-    # sound_root = {}
-    for tag in root.find("SoundRoot") or []:
-        pass  # Dict of Sound by name
+def parse_string(string: str) -> Union[bool, int, float, tuple, Path, str]:
+    string = string.strip()
 
-    # TODO: Build each <ParticleActionList> in <ParticleActionRoot> (11)
-    # particle_action_root = {}
-    for tag in root.find("ParticleActionRoot") or []:
-        pass  # Dict of ParticleActionList by name
+    # Check if string is a boolean
+    if re.match(r"^\s*(?i)(true)\s*$", string):
+        return True
+    elif re.match(r"^\s*(?i)(false)\s*$", string):
+        return False
 
-    # Globals
-    g = root.find("Global")
-    # TODO: Build <CameraPos> and <CaveCameraPos> (12)
-    # story["camera_pos"] = parse_camera_pos(g.find("CameraPos"))
-    # story["cave_camera_pos"] = parse_cave_camera_pos(g.find("CaveCameraPos"))
-    story["background"] = Color(g.find("Background"))
-    story["wand_navigation"] = WandNavigation(g.find("WandNavigation"))
+    # Check if string is an integer
+    if re.match(r"^\s*-?(\d+)\s*$", string):
+        return int(string)
 
-    return story
+    # Check if string is a float
+    if re.match(r"^\s*-?(\d+(\.\d+))\s*$", string):
+        return float(string)
+
+    # Check if string is a tuple (of integers or floats)
+    if re.match(
+        r"\(?\s*-?(\d)\s*,\s*-?(\d)\s*,\s*-?(\d)?\)?",
+        string,
+    ):
+        string = string.strip("()").split(",")
+        return tuple(int(x) for x in string)
+    elif re.match(
+        r"\(?\s*-?(\d+(\.\d+))\s*,\s*-?(\d+(\.\d+))\s*,\s*-?(\d+(\.\d+))?\)?",
+        string,
+    ):
+        string = string.strip("()").split(",")
+        return tuple(float(x) for x in string)
+
+    # Check if string is a path
+    if re.match(r"^[.\/]", string):
+        return Path(string)
+
+    # Plain text
+    return string
+
+
+def parse_attributes(xml: ET.Element) -> dict:
+    return dict(
+        # Use snake_case when val is not a dict
+        (key.replace("-", "_"), parse_string(value))
+        for key, value in xml.attrib.items()
+    )
+
+
+def parse_recursive(xml: ET.Element) -> dict:
+    key = xml.tag
+    val = parse_attributes(xml)
+    title_to_snake = re.compile(r"(?<!^)(?=[A-Z])")
+
+    if val or xml.find("*") is not None:
+        for child in xml:
+            val = val | parse_recursive(child)
+
+        # Only add text property if it isn't empty
+        text = parse_string(xml.text) if xml.text is not None else None
+        if text != "" and text is not None:
+            val["text"] = text
```

Comment:
I think this could be simplified to the following.
```suggestion
        if text:
            val["text"] = text
```

Both empty string and None are falsy